AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/AWSCognitoAuthPlugin+Configure.swift

```diff
@@ -89,6 +91,15 @@ extension AWSCognitoAuthPlugin {
                 frameworkMetadata: AmplifyAWSServiceConfiguration.frameworkMetaData(),
                 region: userPoolConfig.region
             )
+
+            if let provider = self as? UserAgentSuffixProvider, !provider.userAgentSuffix.isEmpty {
```

Comment:
So how does the receiver (`self`) sometimes conform and not to the `UserAgentSuffixProvider` protocol? by subclassing then extending `UserAgentSuffixProvider`? (apologies if I'm missing something here).

Reply:
Oh, you just make the plugin conform to `UserAgentSuffixProvider` :) 

```swift
extension AWSCognitoAuthPlugin: UserAgentSuffixProvider {
    var userAgentSuffix: String {
        "yourCustomSuffix"
    }
}
```

Reply:
Sorry, I was referring to two cases with the same plugin:

1. Used from the context where you **do** want to implement `UserAgentSuffixProvider` (like what you have now)
2. Used from a context where you **do not** want the user agent suffix.

I assumed we wanted to support both cases from the same plugin, but it sounds like we don't. If that's the case, this looks good.

Reply:
I was thinking about this a bit more, and maybe another approach to consider would be:

Making `AWSCognitoAuthPlugin` itself conform to `HttpClientEngine`.

Pros
* No need for an additional semi-public SPI
* Gives us the flexibility to decide at **run-time** to append a suffix or not.

Cons
* Requires `AWSCognitoAuthPlugin` to hold to yet one more reference (in this case, the `HttpClientEngine` which it proxies).
* ?

Reply:
>Used from the context where you **do** want to implement UserAgentSuffixProvider (like what you have now)
>Used from a context where you **do not** want the user agent suffix.

Yes, this approach is either all or nothing, which is OK for what we need right now. The protocol conformance does allow us to avoid having the Proxy set up for **all** requests though. So for a normal use case (i.e. the plugin not conforming to the protocol), nothing would change.

----

>No need for an additional semi-public SPI

We would still need an new external `@_spi` API to let the plugin know what suffix to include in the User-Agent, right?
Or am I missing something?

> Gives us the flexibility to decide at run-time to append a suffix or not.
🤔  Isn't this what we are doing already? I.e. at runtime we check if the plugin conforms to the protocol (either `HttpClientEngine` or `UserAgentSuffixProvider ` and if so we create the proxy.

If you meant making the plugin **always** conform to `HttpClientEngine`, then we would end up creating the proxy for all users (even those who don't need extra User-Agent stuff) and we wanted to avoid that.

Reply:
From what you're saying, and what is missing from this PR, I think I understand your proposal (please correct me if I'm wrong): Some other library or app (not amplify-swift) will contribute this extension in their own library (hoping somebody else isn't doing it also):

```
extension AWSCognitoAuthPlugin: UserAgentSuffixProvider {
...
```

By doing so, that other target decides what suffix to provide. Is this correct?

> We would still need an new external @_spi API to let the plugin know what suffix to include in the User-Agent, right? Or am I missing something?

No, we could do away with `UserAgentSuffixProvider` altogether using the same approach you have here with `HttpClientEngine`.

> 🤔 Isn't this what we are doing already? I.e. at runtime we check if the plugin conforms to the protocol (either HttpClientEngine or UserAgentSuffixProvider  and if so we create the proxy.

Assuming that what you are proposing is what I asked above, the runtime check aspect of it, yes. However, this still leads us with a build-time "hard-coded" implementation of `UserAgentSuffixProvider` (weather done from the `AWSCognitoAuthPlugin` target or elsewhere). What I'm trying to see is if it's worth doing a little extra work so that the decision on wether to contribute a suffix can be done dynamically.

<hr>

For the record, I'm not a big fan of using category extensions to append logic (possibly non-deterministically) to a logical unit (e.g. AWSCognitoAuthPlugin) from outside its own target, e.g. leading to `Redundant conformance of protocol` errors where it may not be clear where the original one is or not being able to do anything about it if someone beat you to it.